packages/iTwinUI-react/src/core/ThemeProvider/ThemeProvider.tsx

```diff
@@ -3,31 +3,77 @@
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
 import React from 'react';
-import { ThemeOptions, ThemeType, useTheme } from '../utils';
-
-export type ThemeProviderProps = {
-  /**
-   * Theme to be applied. If not set, light theme will be used.
-   */
-  theme?: ThemeType;
-  /**
-   * Optional children.
-   */
-  children?: React.ReactNode;
-  /**
-   * Options that can be specified to override default theming behavior.
-   */
-  themeOptions?: ThemeOptions;
-};
+import cx from 'classnames';
+import {
+  PolymorphicComponentProps,
+  PolymorphicForwardRefComponent,
+  useMediaQuery,
+} from '../utils';
+import {
+  useTheme,
+  ThemeOptions,
+  ThemeType,
+  UseThemeProps,
+} from '../utils/hooks/useTheme';
+
+export type ThemeProviderProps<
+  T extends React.ElementType = 'div'
+> = PolymorphicComponentProps<T, UseThemeProps>;
 
 /**
  * Component providing global styles that are required for all components and allows changing theme.
  */
-export const ThemeProvider = (props: ThemeProviderProps) => {
-  const { theme, children, themeOptions } = props;
+export const ThemeProvider = React.forwardRef((props, ref) => {
+  const {
+    theme,
+    children,
+    themeOptions,
+    as: Element = 'div',
+    className,
+    ...rest
+  } = props;
 
-  useTheme(theme, themeOptions);
-  return <>{children}</>;
-};
+  const hasChildren = React.Children.count(children) > 0;
+  const parentContext = React.useContext(ThemeContext);
+  const prefersDark = useMediaQuery('(prefers-color-scheme: dark)');
+  const prefersHighContrast = useMediaQuery('(prefers-contrast: more)');
+
+  const shouldApplyDark = theme === 'dark' || (theme === 'os' && prefersDark);
+  const shouldApplyHC = themeOptions?.highContrast ?? prefersHighContrast;
+
+  // only provide context if wrapped around children
+  return hasChildren ? (
+    <ThemeContext.Provider value={{ theme, themeOptions }}>
+      <Element
```

Comment:
I made this polymorphic because a `<div>` might not always be desirable.

The user will now be able to do this for better SSR support:
```jsx
<ThemeProvider as='body'>
  <App />
</ThemeProvider>
```